File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_11/BlockHoundTest.java
package com.fResult.reactor.ch5_11;

import java.util.ArrayList;
import java.util.ServiceLoader;
import java.util.concurrent.atomic.AtomicBoolean;
import lombok.SneakyThrows;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import reactor.blockhound.BlockHound;
import reactor.blockhound.integration.BlockHoundIntegration;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.test.StepVerifier;

@Log4j2
class BlockHoundTest {
  private static final AtomicBoolean BLOCK_HOUND = new AtomicBoolean();

  private static class BlockingCallError extends Error {
    public BlockingCallError(String msg) {
      super(msg);
    }
  }

  @BeforeEach
  void setUp() {
    BLOCK_HOUND.set(true);
    val integrations = new ArrayList<BlockHoundIntegration>();
    val services = ServiceLoader.load(BlockHoundIntegration.class);
    services.forEach(integrations::add);

    integrations.add(
        builder -> {
          builder.blockingMethodCallback(
              blockingMethod -> {
                if (BLOCK_HOUND.get()) {
                  log.error("Blocking method detected at: {}", blockingMethod);

                  throw new BlockingCallError(blockingMethod.toString());
                }
              });
        });

    BlockHound.install(integrations.toArray(BlockHoundIntegration[]::new));
  }

  @Test
  void ok() {
    StepVerifier.create(buildBlockingMono().subscribeOn(Schedulers.boundedElastic()))
        .expectNext(1L)
        .verifyComplete();
  }

  @Test
  void notOk() {
    StepVerifier.create(buildBlockingMono().subscribeOn(Schedulers.parallel()))
        .expectErrorMatches(this::isBlockingCallError)
        .verify();
  }

  @AfterEach
  void tearDown() {
    BLOCK_HOUND.set(false);
  }

  private Mono<Long> buildBlockingMono() {
    return Mono.just(1L).doOnNext(ignored -> block());
  }

  @SneakyThrows(InterruptedException.class)
  private void block() {
    Thread.sleep(1000);
  }

  private boolean isBlockingCallError(Throwable ex) {
    return ex instanceof BlockingCallError;
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_11/BlockHoundTest.kt
package com.fResult.reactor.ch5_11

import org.apache.logging.log4j.LogManager
import reactor.blockhound.BlockHound
import reactor.blockhound.integration.BlockHoundIntegration
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.test.StepVerifier
import java.util.ServiceLoader
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test

class BlockHoundTest {
  private val log = LogManager.getLogger(BlockHoundTest::class.java)

  companion object {
    private val BLOCK_HOUND = AtomicBoolean()

    private class BlockingCallError(msg: String) : Error(msg)
  }

  @BeforeTest
  fun setUp() {
    log.info("BlockHound test started, blocking calls are now prohibited.")
    BLOCK_HOUND.set(true)
    val integrations = mutableListOf<BlockHoundIntegration>()
    val services = ServiceLoader.load(BlockHoundIntegration::class.java)
    services.forEach(integrations::add)

    integrations.add { builder ->
      builder.blockingMethodCallback { blockingMethod ->
        if (BLOCK_HOUND.get()) {
          log.error("Blocking call detected at: $blockingMethod")

          throw BlockingCallError(blockingMethod.toString())
        }
      }
    }

    BlockHound.install(*integrations.toTypedArray());
  }

  @AfterTest
  fun after() {
    BLOCK_HOUND.set(false)
    log.info("BlockHound test completed, blocking calls are now allowed.")
  }

  @Test
  fun ok() {
    StepVerifier.create(buildBlockingMono().subscribeOn(Schedulers.boundedElastic()))
      .expectNext(1L)
      .verifyComplete()
  }

  @Test
  fun notOk() {
    StepVerifier.create(buildBlockingMono().subscribeOn(Schedulers.parallel()))
      .expectErrorMatches(::isBlockingCallError)
      .verify()
  }

  private fun buildBlockingMono() = Mono.just(1L).doOnNext { block() }

  private fun block() = Thread.sleep(1000)

  private fun isBlockingCallError(ex: Throwable) = ex is BlockingCallError
}
